src/script/auth/page/OAuthPermissions.tsx

```diff
@@ -0,0 +1,258 @@
+import {AssetRemoteData} from 'src/script/assets/AssetRemoteData';
+import {AssetRepository} from 'src/script/assets/AssetRepository';
+import {handleKeyDown, KEY} from 'Util/KeyboardUtil';
+import {loadDataUrl} from 'Util/util';
+
+import {
+  boxCSS,
+  buttonCSS,
+  buttonsCSS,
+  containerCSS,
+  headerCSS,
+  listCSS,
+  teamImageCSS,
+  textCSS,
+} from './OauthPermissions.styles';
+import {Page} from './Page';
+
+import {Config} from '../../Config';
+import {oauthStrings} from '../../strings';
+import {actionRoot} from '../module/action';
+import {bindActionCreators, RootState} from '../module/reducer';
+import * as SelfSelector from '../module/selector/SelfSelector';
+import {oAuthParams, oAuthScope} from '../util/oauthUtil';
+
+interface Props extends React.HTMLProps<HTMLDivElement> {
+  assetRepository?: AssetRepository;
+}
+
+export enum Scope {
+  WRITE_CONVERSATIONS = 'write:conversations',
+  WRITE_CONVERSATIONS_CODE = 'write:conversations_code',
+  READ_SELF = 'read:self',
+  READ_FEATURE_CONFIGS = 'read:feature_configs',
+}
+
+const OAuthPermissionsComponent = ({
+  doLogout,
+  getOAuthApp,
+  selfUser,
+  selfTeamId,
+  assetRepository = container.resolve(AssetRepository),
+  postOauthCode,
+  getSelf,
+  getTeam,
+}: Props & ConnectedProps & DispatchProps) => {
+  const {formatMessage: _} = useIntl();
+  const [teamImage, setTeamImage] = React.useState<string | ArrayBuffer | undefined>(undefined);
+
+  const [oAuthApp, setOAuthApp] = useState<OAuthClient | null>(null);
+  const oauthParams = oAuthParams(window.location);
+  const oauthScope = oAuthScope(oauthParams);
+
+  const onContinue = async () => {
+    try {
+      const url = await postOauthCode(oauthParams);
+      window.location.replace(url);
+    } catch (error) {
+      console.error(error);
+    }
+  };
+
+  const onCancel = () => {
+    window.location.replace(`${oauthParams.redirect_uri}?error=access_denied`);
+  };
+
+  React.useEffect(() => {
+    const getUserData = async () => {
+      await getSelf();
+      if (selfTeamId) {
+        const team = await getTeam(selfTeamId);
+        const teamIcon = AssetRemoteData.v3(team.icon, selfUser.qualified_id?.domain);
+        if (teamIcon.identifier === 'default') {
+          setTeamImage(`${Config.getConfig().APP_BASE}/image/logo/wire-logo-120.png`);
+        } else {
+          const teamImageBlob = await assetRepository.load(teamIcon);
+          setTeamImage(teamImageBlob && (await loadDataUrl(teamImageBlob)));
+        }
+      }
+      setOAuthApp(!!oauthParams.client_id ? await getOAuthApp(oauthParams.client_id) : null);
+    };
+    getUserData().catch(error => {
+      console.error(error);
+    });
+  }, [
+    assetRepository,
+    getOAuthApp,
+    getSelf,
+    getTeam,
+    oauthParams.client_id,
+    selfTeamId,
+    selfUser.qualified_id?.domain,
+  ]);
+
+  return (
+    <Page>
+      <ContainerXS centerText verticalCenter css={containerCSS}>
+        {!oAuthApp ? (
```

Comment:
Wouldn't it be better to use a isLoading variable? When we fail to get the oAuthApp for whatever reason, we show the loader indefinitely

Reply:
i tried just now with the redux `isFetching` state for the auth Selector. Unfortunately because multiple fetching calls are going off in sequence it creates a bit of a jump. 

Since the oAuthApp is the last thing we are loading this works slightly nicer. 

i have added a logout if the function errors to prevent an infinite loading case.